### PR TITLE
autocreate labels to supercat mapping, with user-input also allowed

### PR DESCRIPTION
**Pull request by aliu22**
create a mapping function from labels to supercategories, allowing for user override
supercat_automation_2

### CODE REVIEW

File: datasets.py [hunk 1]
```diff
@@ -144,7 +232,8 @@ def __init__(self, transform):
 
 
         # Maps each label to number of supercategory group, which is listed in keys of GROUPINGS_TO_NAMES (optional)
-        self.group_mapping = None
+        # self.group_mapping = None
```
Comment (Angelina-Wang) [inline datasets.py]: delete this line

File: datasets.py [hunk 2]
```diff
@@ -117,6 +122,89 @@ def read_xml_content(xml_file):
...
+        # create list of supercats for easy indexing & ordering
+        supercat_list = list(supercategories_to_names.keys())
+        # score array for each supercat, for softmax calculation
+        score_arr = []
+
+        for supercat in (supercat_list):
+            word1 = nlp(supercategories_to_names[supercat])
+            word2 = nlp(word)
+            cur_score = word1.similarity(word2)
+            if cur_score is None:
```
Comment (Angelina-Wang) [inline datasets.py]: when does this happen?

File: datasets.py [hunk 3]
```diff
@@ -117,6 +122,89 @@ def read_xml_content(xml_file):
...
+    for label, hr_label in labels_to_names.items():
+        if override_map and hr_label in override_map:
+            # skip if user is overriding with override_map
+            continue
+        # retrieve tuple result from dist_calculator
+        supercat_match = dist_calculator(hr_label)
+        result_label_to_group_map[label] = supercat_match
+        
+    # sort from least to most confident based on softmax scores
+    result_label_to_group_map = dict(sorted(result_label_to_group_map.items(), key=lambda item: item[1][1]))
```
Comment (Angelina-Wang) [inline datasets.py]: I think you mean OrderedDict here and not dict?

File: datasets.py [hunk 4]
```diff
@@ -117,6 +122,89 @@ def read_xml_content(xml_file):
...
+            supercat = names_to_supercategories.get(v)
+            assert label in labels_to_names and supercat in supercategories_to_names, "override string not valid"
+            override_nonhuman_readable[label] = supercat
+        # add overriden dict to result dict 
+        result_label_to_group_map.update(override_nonhuman_readable)
+
+    # print mapping in human-readable form so user can adjust if necessary
+    print("Ranked from least to most confident, here are the \nlabel->supercategory mappings:\n(change as necessary using override_map)")
+    print("-------------------------------")
+    for k, v in result_label_to_group_map.items():
```
Comment (Angelina-Wang) [inline datasets.py]: does this print all of them? maybe just do the first 20 or something, and mention that in the comment

File: datasets.py [hunk 5]
```diff
@@ -117,6 +122,89 @@ def read_xml_content(xml_file):
     11: 'indoor'
 }
 
+def group_mapping_creator(labels_to_names, supercategories_to_names,
```
Comment (Angelina-Wang) [inline datasets.py]: make the default for supercategories_to_names if none is provided, GROUPINGS_TO_NAMES

File: datasets.py (hunk 2 above)
Comment (aliu22) [reply datasets.py]: Thanks -- I think you're referring to the "cur_score is None" check. I took a look at the [documentation](https://spacy.io/api/doc#similarity) and the similarity() function always returns a float, so I removed the None check. (Let me know if this is what you were referring to?)


Also, unrelated I think but in line 152, I'm needlessly computing the nlp tokens for the supercats each time the dist_calculator() function is being run, so I precomputed the tokens outside of the function.

File: datasets.py (hunk 3 above)
Comment (aliu22) [reply datasets.py]: changed, thanks!

File: datasets.py (hunk 4 above)
Comment (aliu22) [reply datasets.py]: done, printing the 20 least confident now

File: datasets.py (hunk 5 above)
Comment (aliu22) [reply datasets.py]: done, thanks.

I'm looking into cleaning up the larger GROUPINGS_TO_NAMES solution you mentioned now

File: Gender Analysis.ipynb [hunk 6]
```diff
@@ -308,7 +308,7 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "data_folder = None\n",
+    "data_folder = None # \"revise-tool/data/coco/2014data/train2014/\"\n",
```
Comment (Angelina-Wang) [inline Gender Analysis.ipynb]: please remove comment

File: Object Analysis.ipynb [hunk 7]
```diff
@@ -290,7 +290,7 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "data_folder = None\n",
+    "data_folder = None #revise-tool/data/coco/2014data/train2014/\n",
```
Comment (Angelina-Wang) [inline Object Analysis.ipynb]: same, please remove

File: datasets.py [hunk 8]
```diff
@@ -117,6 +122,84 @@ def read_xml_content(xml_file):
...
+        for k,v in override_map.items():
+            label = names_to_labels.get(k)
+            supercat = names_to_supercategories.get(v)
+            assert label in labels_to_names and supercat in supercategories_to_names, "override string not valid"
+            override_nonhuman_readable[label] = supercat
+        # add overriden dict to result dict 
+        result_label_to_group_map.update(override_nonhuman_readable)
+
+    # print mapping in human-readable form so user can adjust if necessary
+    print("Here are the 20 LEAST confident labels to supercategory mappings, ranked in increasing confidence.\nChange as necessary using override_map)")
```
Comment (Angelina-Wang) [inline datasets.py]: maybe change phrasing to be something like "here are 20 of the least confident...." so it doesn't seem like this specific 20 was picked for a reason, when this number was actually arbitrary

File: datasets.py [hunk 9]
```diff
@@ -583,6 +673,7 @@ class YfccPlacesDataset(data.Dataset):
     def __init__(self, transform, metric=0):
         self.transform = transform
         
+        self.supercategories_to_names = DEFAULT_GROUPINGS_TO_NAMES
```
Comment (Angelina-Wang) [inline datasets.py]: for these that you added to Imagenet and yfccplaces, I believe they also will need to be accompanied with a "self.group_mapping = group_mapping_creator(self.labels_to_names, self.supercategories_to_names)"?

File: object_based.py [hunk 10]
```diff
@@ -47,17 +47,17 @@ def count_cooccurrence(dataloader, args):
         for j in range(i+1, len(all_categories)):
             counts["{0}-{1}".format(i, j)] = 0
 
-    groupings_size = [[] for i in range(len(GROUPINGS_TO_NAMES))]
-    groupings_dist = [[] for i in range(len(GROUPINGS_TO_NAMES))]
+    groupings_size = [[] for i in range(len(DEFAULT_GROUPINGS_TO_NAMES))]
```
Comment (Angelina-Wang) [inline object_based.py]: instead of DEFAULT_GROUPINGS_TO_NAMES throughout these files, it should say dataloader.dataset.supercategories_to_names

**Pull request merged**